imgix/urlbuilder.py

```diff
@@ -140,35 +140,25 @@ def create_srcset(self, path, params=None):
         if not params:
             params = {}
 
-        width = params['w'] if 'w' in params else None
-        height = params['h'] if 'h' in params else None
-        aspect_ratio = params['ar'] if 'ar' in params else None
+        width = 'w' in params
+        height = 'h' in params
+        aspect_ratio = 'ar' in params
 
         if (width or (height and aspect_ratio)):
             return self._build_srcset_DPR(path, params)
         else:
             return self._build_srcset_pairs(path, params)
 
     def _build_srcset_pairs(self, path, params):
-        srcset = ''
-
-        for i in range(len(SRCSET_TARGET_WIDTHS)):
-            current_width = SRCSET_TARGET_WIDTHS[i]
-            current_params = params
-            current_params['w'] = current_width
-            srcset += self.create_url(path, current_params) \
-                + ' ' + str(current_width) + 'w,\n'
-
-        return srcset[0:-2]
+        srcset_parts = [
+            self.create_url(path, dict(params, w=w)) + " " + str(w) + "w"
```

Comment:
I'm a little concerned about creating a new copy of `params` here for every width (same thing goes for L161). Perhaps it would be more efficient to replace the value of `w` each loop, the same as before, and then restore it after `srcset_parts` has completed running.

Reply:
I am not concerned about this at all. I approach this kind of question with a measuring mindset. Unless you have numbers showing that there is a performance issue, I favour code that is easier to understand for humans. This is what I'd call a micro-optimisation and it would only matter in hot loops.

Reply:
I appreciate the focus on making code easier to understand, and in general I would agree with that call. However, we know from past experiences that for most of our users, performance matters a lot for `create_srcset` (same goes for the `create_url` function). These two functions are often called a large number of times for each individual page serve, which can have a dramatic effect on page speeds. For better or worse, we've learned that sacrificing code comprehensibility for better performance is important for these kinds of functions. As much as I would like to have both, I have to insist in this case that we focus on optimizing.

Reply:
Alright, let's focus on optimising. Do you have numbers proving that your suggestion is faster? I honestly have no idea what's faster.

For what it's worth, running `%timeit UrlBuilder('example.com').create_srcset('foo')` with the version in master and the version in this PR holds a similar result on my computer (about 740µs per loop)

Reply:
@rik Sorry for the hold up, I had to turn my attention to some other stuff this week. 
I profiled the memory usage on `_build_srcset_pairs` in this PR versus the version in master. Here are some numbers when running `create_srcset()` with no imgix parameters:
```
# pull request
>>> ub.create_srcset('image.jpg')
Partition of a set of 36 objects. Total size = 4251 bytes.
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0     31  86     3347  79      3347  79 str
     1      1   3      416  10      3763  89 types.FrameType
     2      1   3      344   8      4107  97 list
     3      1   3       80   2      4187  98 builtins.weakref
     4      2   6       64   2      4251 100 int

===========================================================================

# master
>>> ub.create_srcset('image.jpg')
Partition of a set of 5 objects. Total size = 2561 bytes.
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0      1  20     2001  78      2001  78 str
     1      1  20      416  16      2417  94 types.FrameType
     2      1  20       80   3      2497  98 builtins.weakref
     3      2  40       64   2      2561 100 int
```
As you can see, the version in this PR is allocating ~66% more memory. I included this example because many times users will generate a default `srcset` with no additions parameters. Here is another example passing in a dict of imgix parameters:
```
# pull request
ub.create_srcset('image.jpg', {'mono':'red', 'txt':'title', 'txt-color':'#000', 'txt-font':'Arial-Black', 'txt-size':62, 'txt-pad':180})
Partition of a set of 36 objects. Total size = 6762 bytes.
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0     31  86     5858  87      5858  87 str
     1      1   3      416   6      6274  93 types.FrameType
     2      1   3      344   5      6618  98 list
     3      1   3       80   1      6698  99 builtins.weakref
     4      2   6       64   1      6762 100 int

===========================================================================

>>> ub.create_srcset('image.jpg', {'mono':'red', 'txt':'title', 'txt-color':'#000', 'txt-font':'Arial-Black', 'txt-size':62, 'txt-pad':180})
Partition of a set of 5 objects. Total size = 5072 bytes.
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0      1  20     4512  89      4512  89 str
     1      1  20      416   8      4928  97 types.FrameType
     2      1  20       80   2      5008  99 builtins.weakref
     3      2  40       64   1      5072 100 int
```
Similarly, the version in this PR reflects higher memory usage, allocating ~33% more. This, coupled with the similar run times as you pointed out, indicates that we need to rethink how we want to refactor these two methods.

Reply:
Interesting. What if you change `srcset_parts` from a list comprehension to a generator expression?

Reply:
I've never profiled Python memory. If you could share how you profiled this, I'm happy to try a few different things and see what has the lowest memory profile.

Reply:
Hmm I'm not sure, it sounds like it would yield better results than a list comprehension. Feel free to give it a try, here's what I did to profile the code:
```py
from guppy import hpy
...
def _build_srcset_pairs(self, path, params):
        h = hpy()
        h.setrelheap()
        srcset = ''

        for i in range(len(SRCSET_TARGET_WIDTHS)):
            current_width = SRCSET_TARGET_WIDTHS[i]
            current_params = params
            current_params['w'] = current_width
            srcset += self.create_url(path, current_params) \
                + ' ' + str(current_width) + 'w,\n'

        print(h.heap())
        return srcset[0:-2]
```

Reply:
Hey @rik, any updates? Would love to have this merged if we can reach a consensus.